include/dlaf/communication/kernels/p2p.h

```diff
@@ -0,0 +1,70 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// @file
+
+#include <mpi.h>
+
+#include "dlaf/common/callable_object.h"
+#include "dlaf/common/pipeline.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/communication/message.h"
+#include "dlaf/matrix/tile.h"
+#include "dlaf/sender/traits.h"
+#include "dlaf/sender/transform_mpi.h"
+#include "dlaf/sender/when_all_lift.h"
+
+namespace dlaf::comm {
+
+namespace internal {
+
+// Non-blocking point to point send
+template <class T, Device D>
+void send(const matrix::Tile<const T, D>& tile, IndexT_MPI dest, IndexT_MPI tag,
+          common::PromiseGuard<Communicator> pcomm, MPI_Request* req) {
+  auto msg = comm::make_message(common::make_data(tile));
+  DLAF_MPI_CHECK_ERROR(
+      MPI_Isend(const_cast<T*>(msg.data()), msg.count(), msg.mpi_type(), dest, tag, pcomm.ref(), req));
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(send);
+
+// Non-blocking point to point receive
+template <class T, Device D>
+auto recv(const matrix::Tile<T, D>& tile, IndexT_MPI source, IndexT_MPI tag,
+          common::PromiseGuard<Communicator> pcomm, MPI_Request* req) {
+  auto msg = comm::make_message(common::make_data(tile));
+  DLAF_MPI_CHECK_ERROR(
+      MPI_Irecv(msg.data(), msg.count(), msg.mpi_type(), source, tag, pcomm.ref(), req));
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(recv);
+}
+
+template <class CommSender, class Sender>
+void scheduleSend(IndexT_MPI dest, CommSender&& pcomm, IndexT_MPI tag, Sender&& tile) {
```

Comment:
How do you plan to use it if communications are not ordered?
with a shared_future like sender?

Reply:
If you are talking about the problem that currently we implicitly require in our internal communication kernels to get a `PromiseGuard` instead of the actual `Communicator`, I agree with you and it is something I'd like to discuss with @msimberg in these days.

On the contrary, if it is not that, I'd ask you to explain a bit more in detail the problem you are thinking about.

Reply:
@rasolca confirmed to me that the point was exactly that one. This is going to be addressed in #573.